src/test/java/winstone/HttpConnectorFactoryTest.java

```diff
@@ -49,6 +57,20 @@ public void testListenAddress() throws Exception {
 
     @Test
     public void writePortInFile() throws Exception {
+        Callable<String> parallelVerifier = () -> {
+            System.out.println("Async Verifier started!");
+            Path portFile = Paths.get(tmp.getRoot().getAbsolutePath(), "subdir/jenkins.port");
+
+            Awaitility.await()
+                    .pollInterval(1, TimeUnit.MICROSECONDS)
+                    .atMost(5, TimeUnit.SECONDS)
+                    .until(() -> Files.exists(portFile));
+
+            return FileUtils.readFileToString(portFile.toFile(), StandardCharsets.UTF_8);
+        };
+        ExecutorService executorService = Executors.newSingleThreadExecutor();
+        Future<String> futurePort = executorService.submit(parallelVerifier);
```

Comment:
what do we need this callable for?  can't we just await and once it's found the file read it into a string.

i.e. why do we need a future?

From my reading of the docs this doesn't look right, https://github.com/awaitility/awaitility/wiki/Usage

Reply:
This Test tries to imitate that if parallel process trying to access the port the situation when `portFile` already exists, but `port` still not written can't happen, so we are using Future to get first read result after `portFile` detection and we need two threads running in parallel during test execution. I thought that `await` will always block the current thread until condition (I will re-read the doc, maybe I missed something).

Reply:
@timja maybe I'm missing something or we are misunderstanding each other, the goal of test changes is to verify that there is no time gap between `portFile` creation and `port` value write. Already existed asserts are fine and checks port value was written and they works as expected, but they don't check that write-process is atomic. So second Thread is used to hit portFile as often as possible and once file exists read the first value, later in `Future` we could check it is not empty (line 88). I don't see clear way to use  http://www.awaitility.org/ to imitate such scenario instead of second thread. Please correct me if I'm missing something.

Reply:
Ah I see I misunderstood :)

Reply:
> we need two threads running in parallel during test execution

The race in the original code is not between two readers of the file but rather between the writer of the file and a reader of the file. The test as modified in this PR is not actually testing this scenario because it is always reading the file after the writer is finished.

To make the test actually test the scenario that failed in production you would need to do this:

```java
diff --git a/src/main/java/winstone/Launcher.java b/src/main/java/winstone/Launcher.java
index 0a8485b..bc3dba6 100644
--- a/src/main/java/winstone/Launcher.java
+++ b/src/main/java/winstone/Launcher.java
@@ -230,6 +230,11 @@ public class Launcher implements Runnable {
                     Path portDir = portFile.getParent();
                     Files.createDirectories(portDir);
                     try (BufferedWriter writer = Files.newBufferedWriter(portFile, StandardCharsets.UTF_8)) {
+                        try {
+                            Thread.sleep(5000L);
+                        } catch (InterruptedException e) {
+                            // ignore
+                        }
                         writer.write(Integer.toString(port));
                     }
                 } else {
@@ -49,6 +58,7 @@ public class HttpConnectorFactoryTest extends AbstractWinstoneTest {
 
     @Test
     public void writePortInFile() throws Exception {
+        Future<Integer> futurePort = Executors.newSingleThreadExecutor().submit(() -> {
             Map<String,String> args = new HashMap<>();
             args.put("warfile", "target/test-classes/test.war");
             args.put("prefix", "/");
@@ -58,13 +68,20 @@ public class HttpConnectorFactoryTest extends AbstractWinstoneTest {
             System.setProperty(WINSTONE_PORT_FILE_NAME_PROPERTY, portFile.toString());
             try {
                 winstone = new Launcher(args);
-            int port = ((ServerConnector) winstone.server.getConnectors()[0]).getLocalPort();
-            try (BufferedReader reader = Files.newBufferedReader(portFile, StandardCharsets.UTF_8)) {
-                String portInFile = reader.readLine();
-                assertEquals(Integer.toString(port), portInFile);
-            }
+                return ((ServerConnector) winstone.server.getConnectors()[0]).getLocalPort();
             } finally {
                 System.clearProperty(WINSTONE_PORT_FILE_NAME_PROPERTY);
             }
+        });
+        Path portFile = Paths.get(tmp.getRoot().getAbsolutePath(), "subdir/jenkins.port");
+
+        Awaitility.await()
+                .pollInterval(1, TimeUnit.MICROSECONDS)
+                .atMost(5, TimeUnit.SECONDS)
+                .until(() -> Files.exists(portFile));
+
+        String portFileString = Files.readString(portFile, StandardCharsets.UTF_8);
+        assertEquals(Integer.toString(futurePort.get()), portFileString);
+        assertFalse("Port value should not be empty in any time", portFileString.isEmpty());
     }
 }
```

This adds a five-second sleep after truncating the file but before writing the value to expose the race. Then, from the test, we start the server asynchronously. The server truncates the file and starts sleeping. Now from the test we try to read the port and find that it is empty because the server has not written it yet (the server is still sleeping). This reproduces the original problem.

Obviously the five-second sleep should not be checked in. But if we are to change the test at all, we should change it as described above rather than as implemented in this PR. Without the sleep, the above change will result in the test failing as often as it did in production: rarely, but more often on systems with slow I/O.

Reply:
I made changes based on your idea, please see [586feef](https://github.com/jenkinsci/winstone/pull/249/commits/586feeff71811269c7bc802b44b0b2fb760a4cfb)